src/util/taskManager.test.js

```diff
@@ -0,0 +1,116 @@
+// @flow
+
+import {TaskManager} from "./taskManager";
+
+import {SilentTaskReporter} from "./taskReporter";
+
+describe("util/taskManager", () => {
+  class TestCase {
+    taskManager: TaskManager;
+    reporter: SilentTaskReporter;
+    constructor() {
+      this.reporter = new SilentTaskReporter();
+      this.taskManager = new TaskManager(this.reporter);
+    }
+    start(task: string): TaskManager {
+      return this.taskManager.start(task);
+    }
+    finish(task: string) {
+      this.taskManager.finish(task);
+      return this;
+    }
+  }
+  it("errors when finishing an unregistered task", () => {
+    const fail = () => new TestCase().finish("foo");
+    expect(fail).toThrowError("Task foo not registered");
+  });
+  it("errors when starting a a task twice", () => {
+    const testCase = new TestCase();
+    testCase.start("foo");
+    const fail = () => testCase.start("foo");
+    expect(fail).toThrow("Task foo already registered");
+  });
+  it("errors when finishing a task twice", () => {
+    const testCase = new TestCase();
+    testCase.start("foo");
+    testCase.finish("foo");
+    const fail = () => testCase.finish("foo");
+    expect(fail).toThrow("Task foo not registered");
+  });
+  it("works when a parent task is started with a child and finishes, terminating both tasks", () => {
+    const testCase = new TestCase();
+    const fooScope = testCase.start("foo");
+    expect(testCase.reporter.entries()).toEqual([
+      {"taskId": "foo", "type": "START"},
+    ]);
+    fooScope.start("test");
+    testCase.finish("foo");
+    expect(testCase.reporter.entries()).toEqual([
+      {"taskId": "foo", "type": "START"},
+      {"taskId": "foo: test", "type": "START"},
+      {"taskId": "foo: test", "type": "FINISH"},
+      {"taskId": "foo", "type": "FINISH"},
+    ]);
+  });
+  it("works when a child task starts and finishes", () => {
+    const testCase = new TestCase();
+    const fooScope = testCase.start("foo");
+    expect(testCase.reporter.entries()).toEqual([
+      {"taskId": "foo", "type": "START"},
+    ]);
+    fooScope.start("test");
+    fooScope.finish("test");
+    expect(testCase.reporter.entries()).toEqual([
+      {"taskId": "foo", "type": "START"},
+      {"taskId": "foo: test", "type": "START"},
+      {"taskId": "foo: test", "type": "FINISH"},
+    ]);
+  });
```

Comment:
just to be 100% sure the recursion is working, it would be nice to have a double-nested task and check that finishing root finishes the nested task.

Reply:
@decentralion that is tested here I believe:
https://github.com/sourcecred/sourcecred/blob/71c1634861758b49fea97bcfc4fa4542978aaf4a/src/util/taskManager.test.js#L40-L54